### PR TITLE
service.instance.id implementation

### PR DESCRIPTION
**Pull request by zeitlinger**
Implements [service.instance.id spec PR](https://github.com/open-telemetry/semantic-conventions/pull/312)

### CODE REVIEW

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:25]: I don't necessarily disagree but interested in hearing your explanation.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: What do you think about changing this to be implemented via `ResourceConfiguration#createEnvironmentResource`? The behavior here is dependent on whether `service.instance.id` is already included via `OTEL_RESOURCE_ATTRIBUTES` and `createEnvironmentResource` already parses that.

The benefit to splitting it out is that this can be independently disabled. But if its split out, then we really want to make sure it runs _after_ `EnvironmentResourceProvider` and so it should return a higher value for `order()`.

Comment (zeitlinger) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: But I don't have a strong opinion either way.

Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: We have two options: 

- Generate `service.instance.id` first (or early)
- Generate `service.instnace.id` last (or late)

If we generate early, resource providers like `EnvironemntResourceProvider` still can override when a user sets `OTEL_RESOURCE_ATTRIBUTES=1234`. But consider a resource provider which runs after a uuid `service.instance.id` is generated but before `EnvironmentResourceProvider` - it will be able to see that `service.instance.id` has been generated, but won't know how high of a priority that id is. In contrast, if we generate a uuid `service.instance.id` last after all other resource providers have run and only if no `service.instance.id` has already bene provided, then any other resource provider which aims to provide a more useful `service.instance.id` can do so after confirming that no previous resource provider has already generated one. 

Example:

**Generate `service.instance.id` first**

1. ServiceInstanceIdResourceProvider provides a uuid `service.instance.id` 
2. A custom FooResourceProvider aims to provide a more durable `service.instance.id`. It sees that `service.instance.id` already exists and isn't sure if the value it provides is higher quality. 
3. EnvironmentResourceProvider runs an overrides `service.instance.id` if its included in `OTEL_RESOURCE_ATTRIBUTES`.

**Generate `service.instance.id` last**

1. A custom FooResourceProvider aims to provide a more durable `service.instance.id`. It doesn't see a `service.instance.id`, so it safely provides a `service.instance.id`.
2. EnvironmentResourceProvider runs an overrides `service.instance.id` if its included in `OTEL_RESOURCE_ATTRIBUTES`.
3. ServiceInstanceIdResourceProvider provides a uuid `service.instance.id` if and only if no `service.instance.id` has been previously provided. 

Generating last seems better - less ambiguity for other resource providers that may want to provide `service.instance.id`.

Comment (zeitlinger) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: yes, you're right - this is the behavior I wanted - I'll fix it

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:43]: I think we should run this right after the environment resource provider. So set the priority of this to `Integer.MAX_VALUE`, and environment resource provider to `Integer.MAX_VALUE - 1`. 

Then, the `shouldApply` can simply be:
```
return exiting.getAttributes(SERVICE_INSTANCE_ID) == null
```

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (trask) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: > Generating last seems better - less ambiguity for other resource providers that may want to provide `service.instance.id`.

is this the same way we default `service.name` to `unknown_service:java`?

Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: No - the `unknown_service:java` is present when you call `Resource.getDefault()`, which is the [base on top of which](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ResourceConfiguration.java#L88) all other resource providers are merged. 

The outcome of this decision is interesting:
- A resource provider looking to contribute to `service.name` can't make a decision whether or not to provide it based on the presence of `service.name` alone, since its always present.
- A resource provider CAN reliably check if the present `service.name` is the default by comparing to `Resource.getDefault().getAttribute(SERVICE_NAME)`. 
- This is in contrast to `service.instance.id`. If generated first, then another provider would have no way of identifying that the `service.instance.id` is a randomly generated uuid or some value of higher significance. They can check if it matches the uuid format, but can't know if its random or not.
- Generating `service.instance.id` last does make it hard to remove. Although it is still possible via [AutoConfigurationCustomizer#addResourceCustomizer](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java#L45-L46).

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (zeitlinger) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:42]: @jack-berg done :smile:

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/EnvironmentResourceProvider.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/EnvironmentResourceProvider.java:27]: If you want you could have a package private constant in `ServiceInstanceIdResourceProvider` and reference that here:
```
return ServiceInstanceIdResourceProvider.ORDER - 1;
```

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:26]: No need to make this public - we generally try to avoid public constants. Someone can call `new ServiceInstanceIdResourceProvider().createResource()` to access. If we receive user feedback that this is not enough, we should add `public static Resource get()` (or similarly named). But let's wait for feedback.

**Review by jack-berg — APPROVED** [review]
Couple of nits but looks good to me.

@open-telemetry/java-approvers this PR implements a recent [change merged to semantic conventions](https://github.com/open-telemetry/semantic-conventions/pull/312). It has not bee formally released, and the `service.instance.id` attribute is still marked as experimental. So there's a valid question of whether we should include this attribute by default or not.

If we're nervous about merging this until stable, we could include it in the [sdk extension incubator](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/incubator) project instead. The inclusion of the incubator on the classpath acting as an "opt in" from the user.

**Comment by jpkrohling** [comment]
I agree that a release is desirable before merging this: until it's released, changes are always possible, even though they are unlikely for this specific case.

About it being experimental: I believe this change isn't causing backward compatibility issues, other than possibly for cardinality concerns when it comes to metrics. If possible, I'd love this to be opt-out, as it would help us make that semconv stable within a few releases.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: Let's go ahead and move this to `opentelemetry-sdk-extension-incubator`. The incubator is currently bundled with the otel java agent, so the effect should be the same. But moving it to the incubator is the more conservative / accurate home for now given the compatibility guarantees for the stable autoconfigure artifact.

Comment (zeitlinger) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java:20]: yes, will do :smile:

Comment (zeitlinger) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java]: @jack-berg done :smile:

Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java]: I moved this to `io.opentelemetry.sdk.extension.incubator.resources.*`, but looks good. Thanks!

**Pull request merged**
